[JSC] Add uDFG validation phase

Constellation opened the pull request:
#### a8abeceb624b2e05bf50884a32835c856e1f4a2d
<pre>
[JSC] Add uDFG validation phase
<a href="https://bugs.webkit.org/show_bug.cgi?id=246499">https://bugs.webkit.org/show_bug.cgi?id=246499</a>
rdar://101154136

Reviewed by Mark Lam.

This patch adds uDFG validation phase. This allows graceful failure of the compilation if
we accidentally include some linked data in DFG pipeline by enhancing this phase. And at the
same time, we can detect what is wrong in the given program and improve our uDFG quality
incrementally.
Currently, we add a first restriction: all DFG node should have the same origin to the top level
CodeBlock to make the code easy. This paves the way to the initial version of LinkerIR::GlobalObject.
We can enhance it, but it is postponed since it is not so frequent that we have multiple JSGlobalObject
in one DFG function.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/dfg/DFGValidateUnlinked.cpp: Added.
(JSC::DFG::ValidateUnlinked::run):
(JSC::DFG::ValidateUnlinked::validateNode):
(JSC::DFG::performValidateUnlinked):
* Source/JavaScriptCore/dfg/DFGValidateUnlinked.h: Added.
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/255559@main">https://commits.webkit.org/255559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66179c355184e06c857dc2ded258f1bd280cf24c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2101 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102615 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2104 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30441 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85288 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98555 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79377 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28355 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84231 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36841 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79291 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34643 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27468 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3843 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38513 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40772 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81919 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40430 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37351 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18529 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/dfg/DFGValidateUnlinked.h:
```diff
@@ -0,0 +1,45 @@
+// Check uDFG validity of the given graph.
+// If validation fails, we gracefully stop the compilation as a failure.
+// We make this strict to ensure the error is caught here. And at the same time, support more and more DFG
+// nodes properly in uDFG and relax the restriction. The goal is always passing in this function.
```
/The goal is always passing in this function/The goal is to always pass in this function eventually/

Constellation replied:
Fixed.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGValidateUnlinked.cpp:
```diff
@@ -0,0 +1,88 @@
+    if (globalObject != m_graph.m_codeBlock->globalObject()) {
+        if (UNLIKELY(Options::dumpUnlinkedDFGValidation())) {
+            m_graph.logAssertionFailure(node, __FILE__, __LINE__, WTF_PRETTY_FUNCTION, "Bad GlobalObject");
+            dataLogLn(RawPointer(globalObject), " <-> ", RawPointer(m_graph.m_codeBlock->globalObject()));
```
Why use `<->`?  Why not use `!=`?

Constellation replied:
Either is fine, fixed.

MenloDorian reviewed (approved):
r=me with suggestions.

Constellation committed:
[JSC] Add uDFG validation phase
https://bugs.webkit.org/show_bug.cgi?id=246499
rdar://101154136

Reviewed by Mark Lam.

This patch adds uDFG validation phase. This allows graceful failure of the compilation if
we accidentally include some linked data in DFG pipeline by enhancing this phase. And at the
same time, we can detect what is wrong in the given program and improve our uDFG quality
incrementally.
Currently, we add a first restriction: all DFG node should have the same origin to the top level
CodeBlock to make the code easy. This paves the way to the initial version of LinkerIR::GlobalObject.
We can enhance it, but it is postponed since it is not so frequent that we have multiple JSGlobalObject
in one DFG function.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/dfg/DFGValidateUnlinked.cpp: Added.
(JSC::DFG::ValidateUnlinked::run):
(JSC::DFG::ValidateUnlinked::validateNode):
(JSC::DFG::performValidateUnlinked):
* Source/JavaScriptCore/dfg/DFGValidateUnlinked.h: Added.
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/255559@main
